zebra-chain/src/sapling/shielded_data.rs

```diff
@@ -86,6 +86,15 @@ where
     AnchorV: AnchorVariant + Clone,
 {
     /// The net value of Sapling spend transfers minus output transfers.
+    ///
+    /// [`ShieldedData`] structurally validates this [value balance consensus
+    /// rule](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus):
+    /// "If effectiveVersion = 4 and there are no Spend descriptions or Output
+    /// descriptions, then valueBalanceSapling MUST be 0."
+    ///
+    /// `value_balance` is a field in [`ShieldedData`], which must have at least
+    /// one spend or output in its `transfers` field. If [`ShieldedData`] is `None`
+    /// then the `value_balance` is always serialized as zero.
```

Comment:
Unfortunately we've missed a check during deserialization.

In the `None` case, if `value_balance != 0`, we should `Err(TransactionError::BadBalance)?`:
https://github.com/ZcashFoundation/zebra/blob/db0cdb74ff3e147a7a3717ccb40cfc2ac36260bf/zebra-chain/src/transaction/serialize.rs#L600-L612

So here's the suggestion for this comment, which covers both cases:
```suggestion
    ///
    /// During deserialization, blocks 
    ///
    /// [`ShieldedData`] validates this [value balance consensus rule](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus):
    ///
    /// "If effectiveVersion = 4 and there are no Spend descriptions or Output
    /// descriptions, then valueBalanceSapling MUST be 0."
    ///
    /// During deserialization, this rule is checked when there are no spends and
    /// no outputs.
    ///
    /// During serialization, this rule is structurally validated by [`ShieldedData`].
    /// `value_balance` is a field in [`ShieldedData`], which must have at least
    /// one spend or output in its `transfers` field. If [`ShieldedData`] is `None`
    /// then there can not possibly be any spends or outputs, and the
    /// `value_balance` is always serialized as zero.
```